odoo/orm/fields.py

```diff
@@ -1225,12 +1228,14 @@ def _value_to_column(v):
             if (operator == 'in') == null_in_condition:
                 # field in {val, False} => field IN vals OR field IS NULL
                 # field not in {val} => field NOT IN vals OR field IS NULL
+                if not nullable:
+                    return sql or SQL("FALSE")
                 sql_null = SQL("%s IS NULL", sql_field)
                 return SQL("(%s OR %s)", sql, sql_null) if sql else sql_null
 
             elif operator == 'not in' and null_in_condition and not sql:
                 # if we have a base query, null values are already exluded
-                return SQL("%s IS NOT NULL", sql_field)
+                return SQL("%s IS NOT NULL", sql_field) if nullable else SQL("TRUE")
```

Comment:
Is it still happened ? Because your new optimization should cover that case right or I miss something ?

Reply:
It should not happen. But the goal here is to be generic and handle all inputs without relying on that particular optimization. I find it easier to test. It may happen if we have a field type that wouldn't get optimized and wanted to generate that code.